README.md
# Autolink  Plugin

[![Build Status](https://img.shields.io/circleci/project/github/dmarushkin/mattermost-plugin-autolink-with-log/master.svg)](https://circleci.com/gh/dmarushkin/mattermost-plugin-autolink-with-log)
[![Code Coverage](https://img.shields.io/codecov/c/github/dmarushkin/mattermost-plugin-autolink-with-log/master.svg)](https://codecov.io/gh/dmarushkin/mattermost-plugin-autolink-with-log)
[![Release](https://img.shields.io/github/v/release/dmarushkin/mattermost-plugin-autolink-with-log)](https://github.com/dmarushkin/mattermost-plugin-autolink-with-log/releases/latest)
[![HW](https://img.shields.io/github/issues/dmarushkin/mattermost-plugin-autolink-with-log/Up%20For%20Grabs?color=dark%20green&label=Help%20Wanted)](https://github.com/dmarushkin/mattermost-plugin-autolink-with-log/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22Up+For+Grabs%22+label%3A%22Help+Wanted%22)

**Maintainer:** [@levb](https://github.com/levb)
**Co-Maintainer:** [@iomodo](https://github.com/iomodo)

This plugin creates regular expression (regexp) patterns that are reformatted into a Markdown link before the message is saved into the database.

Use it to add custom auto-linking on your Mattermost system, such as adding links to your issue tracker based on the regexp patterns.

![image](https://user-images.githubusercontent.com/13119842/58675221-479ad680-8321-11e9-9ad1-b9c42238734f.png)

*Posting a message containing a Jira issue key..*

![image](https://user-images.githubusercontent.com/13119842/58675165-11f5ed80-8321-11e9-9d41-91088a79a11b.png)

*..automatically links to the corresponding issue in the Jira project*

## Configuration
1. Go to **System Console > Plugins > Management** and click **Enable** to enable the Autolink plugin.
    - If you are running Mattermost v5.11 or earlier, you must first go to the [releases page of this GitHub repository](https://github.com/dmarushkin/mattermost-plugin-autolink-with-log), download the latest release, and upload it to your Mattermost instance [following this documentation](https://docs.mattermost.com/administration/plugins.html#plugin-uploads).

2. Modify your `config.json` file to include the types of regexp patterns you wish to match, under the `PluginSettings`. See below for an example of what this should look like.

**Tip**: There are useful Regular Expression tools online to help test and validate that your formulas are working as expected.  One such tool is [Regex101](https://regex101.com/) . Here is an example Regular Expression to capture a post that includes a [VISA card number](https://regex101.com/r/JGKCTN/1) - which you could then obfuscate with the `Pattern` so people don't accidentally share sensitive info in your channels.

## Usage

Autolinks have 3 parts: a **Pattern** which is a regular expression search pattern utilizing the [Golang regexp library](https://golang.org/pkg/regexp/), a **Template** that gets expanded and an optional **Scope** parameter  to define which team/channel the autolink applies to. You can create variables in the pattern with the syntax `(?P<name>...)` which will then be expanded by the corresponding template.

In the template, a variable is denoted by a substring of the form `$name` or `${name}`, where `name` is a non-empty sequence of letters, digits, and underscores. A purely numeric name like $1 refers to the submatch with the corresponding index. In the $name form, name is taken to be as long as possible: $1x is equivalent to ${1x}, not ${1}x, and, $10 is equivalent to ${10}, not ${1}0. To insert a literal $ in the output, use $$ in the template.

The scope must be either a team (`teamname`) or a team and a channel (`teamname/channelname`). Remember that you must provide the entity name, not the entity display name. Since Direct Messages do not belong to any team, scoped matches will not be autolinked on Direct Messages. If more than one scope is provided, matches in at least one of the scopes will be autolinked.

Below is an example of regexp patterns used for autolinking at https://community.mattermost.com, modified in the `config.json` file:

```json5
"PluginSettings": {
    ...
    "Plugins": {
        "mattermost-autolink-with-log": {
            "links": [
                {
                    "Pattern": "(LHS)",
                    "Template": "[LHS](https://docs.mattermost.com/process/training.html#lhs)",
                    "Scope": ["team/off-topic"]
                },
                {
                    "Pattern": "(RHS)",
                    "Template": "[RHS](https://docs.mattermost.com/process/training.html#rhs)",
                    "Scope": ["team/town-square"]
                },
                {
                    "Pattern": "(?i)(Mana)",
                    "Template": "[Mana](https://docs.mattermost.com/process/training.html#mana)"
                },
                {
                    "Pattern": "(?i)(ESR)",
                    "Template": "[ESR](https://docs.mattermost.com/process/training.html#esr)"
                },
                {
                    "Pattern": "((?P<level>0|1|2|3|4|5)/5)",
                    "Template": "[${level}/5](https://docs.mattermost.com/process/training.html#id8)"
                },
                {
                    "Pattern": "(MM)(-)(?P<jira_id>\\d+)",
                    "Template": "[MM-${jira_id}](https://mattermost.atlassian.net/browse/MM-${jira_id})"
                },
                {
                    "Pattern": "https://pre-release\\.mattermost\\.com/core/pl/(?P<id>[a-zA-Z0-9]+)",
                    "Template": "[<jump to convo>](https://pre-release.mattermost.com/core/pl/${id})"
                },
                {
                    "Pattern": "(https://mattermost\\.atlassian\\.net/browse/)(MM)(-)(?P<jira_id>\\d+)",
                    "Template": "[MM-${jira_id}](https://mattermost.atlassian.net/browse/MM-${jira_id})"
                },
                {
                    "Pattern": "https://github\\.com/mattermost/(?P<repo>.+)/pull/(?P<id>\\d+)",
                    "Template": "[pr-${repo}-${id}](https://github.com/mattermost/${repo}/pull/${id})"
                },
                {
                    "Pattern": "https://github\\.com/mattermost/(?P<repo>.+)/issues/(?P<id>\\d+)",
                    "Template": "[issue-${repo}-${id}](https://github.com/mattermost/${repo}/issues/${id})"
                },
                {
                    "Pattern": "(PLT)(-)(?P<jira_id>\\d+)",
                    "Template": "[PLT-${jira_id}](https://mattermost.atlassian.net/browse/PLT-${jira_id})"
                },
                {
                    "Pattern": "(https://mattermost\\.atlassian\\.net/browse/)(PLT)(-)(?P<jira_id>\\d+)",
                    "Template": "[PLT-${jira_id}](https://mattermost.atlassian.net/browse/PLT-${jira_id})"
                }
            ]
        },
    },
    ...
    "PluginStates": {
        ...
        "mattermost-autolink-with-log": {
            "Enable": true
        },
        ...
    }
},
```

## Examples

1. Autolinking `Ticket ####:text with alphanumberic characters and spaces` to a ticket link. Use:
   - Pattern: `(?i)(ticket )(?P<ticket_id>.+)(:)(?P<ticket_info>.*)`, or if the ticket_id is a number, then `(?i)(ticket )(?P<ticket_id>\d+)(:)(?P<ticket_info>.*)`
   - Template: `[Ticket ${ticket_id}: ${ticket_info}](https://github.com/mattermost/mattermost-server/issues/${ticket_id})`
   - Scope: `["teams/committers"]` (optional)

2. Autolinking a link to a GitHub PR to a format "pr-repo-id". Use:
   - Pattern: `https://github\\.com/mattermost/(?P<repo>.+)/pull/(?P<id>\\d+)`
   - Template: `[pr-${repo}-${id}](https://github.com/mattermost/${repo}/pull/${id})`

3. Using autolinking to create group mentions. Use (note that clicking the resulting at-mention redirects to a broken page):
   - Pattern: `@customgroup*`
   - Template: `[@customgroup]( \\* @user1 @user2 @user3 \\* )`

4. For servers with multiple domains (like community and community-daily on the [public Mattermost Server](https://community.mattermost)), a substitution of absolute conversation links to relative links is recommended to prevent issues in the mobile app. Add one pattern for each domain used:
   - Pattern: `https://community\\.mattermost\\.com/(?P\u003cteamname\u003e(?a-zA-Z0-9]+)/(?P\u003cid\u003e[a-zA-Z0-9]+)`
   - Template: `[<jump to convo>](/${teamname}/pl/${id})/${id})`

Autolink the word Handbook to a internal URL on the private team (called `office`), and a private channel (`staff`) in the public team (called `everyone`).
    - Pattern: `(Handbook)`
    - Template: `[Handbook](http://www.mywebsite.com/private/handbook)`
    - Scope: `["office", "everyone/staff"]`

**You can check your pattern with those Regex Testers:**
- https://regex-golang.appspot.com/,
- https://regex101.com/,
- https://www.regextester.com/.

## Configuration Management
The `/autolink` commands allow the users to easily edit the configurations.

 Commands | Description | Usage
 ---|---|---|
 list | Lists all configured links | `/autolink list`
 list \<*linkref*> | List a specific link which matched the link reference | `/autolink list test`
 test \<*linkref*> test-text | Test a link on the text provided | `/autolink test Visa 4356-7891-2345-1111 -- (4111222233334444)`
 enable \<*linkref*> | Enables the link | `/autolink enable Visa`
 disable \<*linkref*> | Disable the link | `/autolink disable Visa`
 add \<*linkref*> | Creates a new link with the name specified in the command  | `/autolink add Visa`
 delete \<*linkref*> |  Delete the link | `/autolink delete Visa`
 set \<*linkref*> \<*field*> *value* | Sets a link's field to a value <br> *Fields* - <br> <ul><li>Template - Sets the Template field</li><li>Pattern - Sets the Pattern field </li> <li> WordMatch - If true uses the [\b word boundaries](https://www.regular-expressions.info/wordboundaries.html) </li>  <li> LogHits - If true all replaced matches will be logged on serverside </li> <li> ProcessBotPosts - If true applies changes to posts made by bot accounts. </li> <li> Scope - Sets the Scope field (`team` or `team/channel` or a whitespace-separated list thereof) </li> | <br> `/autolink set Visa Pattern (?P<VISA>(?P<part1>4\d{3})[ -]?(?P<part2>\d{4})[ -]?(?P<part3>\d{4})[ -]?(?P<LastFour>[0-9]{4}))` <br><br> `/autolink set Visa Template VISA XXXX-XXXX-XXXX-$LastFour` <br><br> `/autolink set Visa WordMatch true` <br><br> `/autolink set Visa LogHits true` <br><br> `/autolink set Visa ProcessBotPosts true` <br><br> `/autolink set Visa Scope team/townsquare` <br><br>


## Development

This plugin contains a server portion.

Use `make dist` to build distributions of the plugin that you can upload to a Mattermost server.
Use `make check-style` to check the style.
Use `make deploy` to deploy the plugin to your local server.

For additional information on developing plugins, refer to [our plugin developer documentation](https://developers.mattermost.com/extend/plugins/).


CHANGELOG.md
# Changelog

The changelog can be found at https://github.com/dmarushkin/mattermost-plugin-autolink-with-log/releases.


server/api/api_test.go
package api

import (
	"bytes"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/dmarushkin/mattermost-plugin-autolink-with-log/server/autolink"
)

type authorizeAll struct{}

func (authorizeAll) IsAuthorizedAdmin(string) (bool, error) {
	return true, nil
}

type linkStore struct {
	prev       []autolink.Autolink
	saveCalled *bool
	saved      *[]autolink.Autolink
}

func (s *linkStore) GetLinks() []autolink.Autolink {
	return s.prev
}

func (s *linkStore) SaveLinks(links []autolink.Autolink) error {
	*s.saved = links
	*s.saveCalled = true
	return nil
}

func TestSetLink(t *testing.T) {
	for _, tc := range []struct {
		name             string
		method           string
		prevLinks        []autolink.Autolink
		link             autolink.Autolink
		expectSaveCalled bool
		expectSaved      []autolink.Autolink
		expectStatus     int
	}{
		{
			name: "happy simple",
			link: autolink.Autolink{
				Name: "test",
			},
			expectStatus:     http.StatusOK,
			expectSaveCalled: true,
			expectSaved: []autolink.Autolink{{
				Name: "test",
			}},
		},
		{
			name: "add new link",
			link: autolink.Autolink{
				Name:     "test1",
				Pattern:  ".*1",
				Template: "test1",
			},
			prevLinks: []autolink.Autolink{{
				Name:     "test2",
				Pattern:  ".*2",
				Template: "test2",
			}},
			expectStatus:     http.StatusOK,
			expectSaveCalled: true,
			expectSaved: []autolink.Autolink{{
				Name:     "test2",
				Pattern:  ".*2",
				Template: "test2",
			}, {
				Name:     "test1",
				Pattern:  ".*1",
				Template: "test1",
			}},
		}, {
			name: "replace link",
			link: autolink.Autolink{
				Name:     "test2",
				Pattern:  ".*2",
				Template: "new template",
			},
			prevLinks: []autolink.Autolink{{
				Name:     "test1",
				Pattern:  ".*1",
				Template: "test1",
			}, {
				Name:     "test2",
				Pattern:  ".*2",
				Template: "test2",
			}, {
				Name:     "test3",
				Pattern:  ".*3",
				Template: "test3",
			}},
			expectStatus:     http.StatusOK,
			expectSaveCalled: true,
			expectSaved: []autolink.Autolink{{
				Name:     "test1",
				Pattern:  ".*1",
				Template: "test1",
			}, {
				Name:     "test2",
				Pattern:  ".*2",
				Template: "new template",
			}, {
				Name:     "test3",
				Pattern:  ".*3",
				Template: "test3",
			}},
		},
		{
			name: "no change",
			link: autolink.Autolink{
				Name:     "test2",
				Pattern:  ".*2",
				Template: "test2",
			},
			prevLinks: []autolink.Autolink{{
				Name:     "test1",
				Pattern:  ".*1",
				Template: "test1",
			}, {
				Name:     "test2",
				Pattern:  ".*2",
				Template: "test2",
			}},
			expectStatus:     http.StatusNotModified,
			expectSaveCalled: false,
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			var saved []autolink.Autolink
			var saveCalled bool

			h := NewHandler(
				&linkStore{
					prev:       tc.prevLinks,
					saveCalled: &saveCalled,
					saved:      &saved,
				},
				authorizeAll{},
			)

			body, err := json.Marshal(tc.link)
			require.NoError(t, err)

			w := httptest.NewRecorder()
			method := "POST"
			if tc.method != "" {
				method = tc.method
			}
			r, err := http.NewRequest(method, "/api/v1/link", bytes.NewReader(body))
			require.NoError(t, err)

			r.Header.Set("Mattermost-Plugin-ID", "testfrom")
			r.Header.Set("Mattermost-User-ID", "testuser")

			h.ServeHTTP(w, r)
			require.Equal(t, tc.expectStatus, w.Code)
			require.Equal(t, tc.expectSaveCalled, saveCalled)
			require.Equal(t, tc.expectSaved, saved)
		})
	}
}


server/api/api.go
package api

import (
	"encoding/json"
	"net/http"

	"github.com/gorilla/mux"
	"github.com/pkg/errors"

	"github.com/dmarushkin/mattermost-plugin-autolink-with-log/server/autolink"
)

type Store interface {
	GetLinks() []autolink.Autolink
	SaveLinks([]autolink.Autolink) error
}

type Authorization interface {
	IsAuthorizedAdmin(userID string) (bool, error)
}

type Handler struct {
	root          *mux.Router
	store         Store
	authorization Authorization
}

func NewHandler(store Store, authorization Authorization) *Handler {
	h := &Handler{
		store:         store,
		authorization: authorization,
	}

	root := mux.NewRouter()
	api := root.PathPrefix("/api/v1").Subrouter()
	api.Use(h.adminOrPluginRequired)
	api.HandleFunc("/link", h.setLink).Methods("POST")

	api.Handle("{anything:.*}", http.NotFoundHandler())

	h.root = root

	return h
}

func (h *Handler) handleError(w http.ResponseWriter, err error) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusInternalServerError)
	b, _ := json.Marshal(struct {
		Error   string `json:"error"`
		Details string `json:"details"`
	}{
		Error:   "An internal error has occurred. Check app server logs for details.",
		Details: err.Error(),
	})
	_, _ = w.Write(b)
}

func (h *Handler) adminOrPluginRequired(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var err error
		authorized := false
		pluginID := r.Header.Get("Mattermost-Plugin-ID")
		if pluginID != "" {
			// All other plugins are allowed
			authorized = true
		}

		userID := r.Header.Get("Mattermost-User-ID")
		if !authorized && userID != "" {
			authorized, err = h.authorization.IsAuthorizedAdmin(userID)
			if err != nil {
				http.Error(w, "Not authorized", http.StatusUnauthorized)
				return
			}
		}

		if !authorized {
			http.Error(w, "Not authorized", http.StatusUnauthorized)
			return
		}
		next.ServeHTTP(w, r)
	})
}

func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	h.root.ServeHTTP(w, r)
}

func (h *Handler) setLink(w http.ResponseWriter, r *http.Request) {
	var newLink autolink.Autolink
	if err := json.NewDecoder(r.Body).Decode(&newLink); err != nil {
		h.handleError(w, errors.Wrap(err, "unable to decode body"))
		return
	}

	links := h.store.GetLinks()
	found := false
	changed := false
	for i := range links {
		if links[i].Name == newLink.Name || links[i].Pattern == newLink.Pattern {
			if !links[i].Equals(newLink) {
				links[i] = newLink
				changed = true
			}
			found = true
			break
		}
	}
	if !found {
		links = append(h.store.GetLinks(), newLink)
		changed = true
	}
	status := http.StatusNotModified
	if changed {
		if err := h.store.SaveLinks(links); err != nil {
			h.handleError(w, errors.Wrap(err, "unable to save link"))
			return
		}
		status = http.StatusOK
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_, _ = w.Write([]byte(`{"status": "OK"}`))
}


server/autolinkclient/client.go
package autolinkclient

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"

	"github.com/dmarushkin/mattermost-plugin-autolink-with-log/server/autolink"
)

const autolinkPluginID = "mattermost-autolink-with-log"

type PluginAPI interface {
	PluginHTTP(*http.Request) *http.Response
}

type Client struct {
	http.Client
}

type pluginAPIRoundTripper struct {
	api PluginAPI
}

func (p *pluginAPIRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	resp := p.api.PluginHTTP(req)
	if resp == nil {
		return nil, fmt.Errorf("failed to make interplugin request")
	}
	return resp, nil
}

func NewClientPlugin(api PluginAPI) *Client {
	client := &Client{}
	client.Transport = &pluginAPIRoundTripper{api}
	return client
}

func (c *Client) Add(links ...autolink.Autolink) error {
	for _, link := range links {
		linkBytes, err := json.Marshal(link)
		if err != nil {
			return err
		}

		req, err := http.NewRequest("POST", "/"+autolinkPluginID+"/api/v1/link", bytes.NewReader(linkBytes))
		if err != nil {
			return err
		}

		resp, err := c.Do(req)
		if err != nil {
			return err
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			respBody, _ := ioutil.ReadAll(resp.Body)
			return fmt.Errorf("unable to install autolink. Error: %v, %v", resp.StatusCode, string(respBody))
		}
	}

	return nil
}


server/autolinkplugin/command.go
package autolinkplugin

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/mattermost/mattermost-server/v6/model"
	"github.com/mattermost/mattermost-server/v6/plugin"
	"github.com/pkg/errors"

	"github.com/dmarushkin/mattermost-plugin-autolink-with-log/server/autolink"
)

const (
	autolinkCommand         = "/autolink"
	optName                 = "Name"
	optTemplate             = "Template"
	optPattern              = "Pattern"
	optScope                = "Scope"
	optDisabled             = "Disabled"
	optProcessBotPosts      = "ProcessBotPosts"
	optDisableNonWordPrefix = "DisableNonWordPrefix"
	optDisableNonWordSuffix = "DisableNonWordSuffix"
	optWordMatch            = "WordMatch"
	optLogHits              = "LogHits"
)

const helpText = "###### Mattermost Autolink Plugin Administration\n" +
	"<linkref> is either the Name of a link, or its number in the `/autolink list` output. A partial Name can be specified, but some commands require it to be uniquely resolved.\n" +
	"* `/autolink add <name>` - add a new link, named <name>.\n" +
	"* `/autolink delete <linkref>` - delete a link.\n" +
	"* `/autolink disable <linkref>` - disable a link.\n" +
	"* `/autolink enable <linkref>` - enable a link.\n" +
	"* `/autolink list <linkref>` - list a specific link.\n" +
	"* `/autolink list <field> value` - list links whose <field> contains value. Here <field> can be Template or Pattern\n" +
	"* `/autolink list` - list all configured links.\n" +
	"* `/autolink set <linkref> <field> value...` - sets a link's field to a value. The entire command line after <field> is used for the value, unescaped, leading/trailing whitespace trimmed.\n" +
	"* `/autolink test <linkref> test-text...` - test a link on a sample.\n" +
	"\n" +
	"Example:\n" +
	"```\n" +
	"/autolink add Visa\n" +
	"/autolink disable Visa\n" +
	`/autolink set Visa Pattern (?P<VISA>(?P<part1>4\d{3})[ -]?(?P<part2>\d{4})[ -]?(?P<part3>\d{4})[ -]?(?P<LastFour>[0-9]{4}))` + "\n" +
	"/autolink set Visa Template VISA XXXX-XXXX-XXXX-$LastFour\n" +
	"/autolink set Visa WordMatch true\n" +
	"/autolink set Visa LogHits true\n" +
	"/autolink set Visa Scope team/townsquare\n" +
	"/autolink set Visa ProcessBotPosts true\n" +
	"/autolink test Visa 4356-7891-2345-1111 -- (4111222233334444)\n" +
	"/autolink enable Visa\n" +
	"```\n" +
	""

type CommandHandlerFunc func(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse

type CommandHandler struct {
	handlers       map[string]CommandHandlerFunc
	defaultHandler CommandHandlerFunc
}

var autolinkCommandHandler = CommandHandler{
	handlers: map[string]CommandHandlerFunc{
		"help":    executeHelp,
		"list":    executeList,
		"delete":  executeDelete,
		"disable": executeDisable,
		"enable":  executeEnable,
		"add":     executeAdd,
		"set":     executeSet,
		"test":    executeTest,
	},
	defaultHandler: executeHelp,
}

func (ch CommandHandler) Handle(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	for n := len(args); n > 0; n-- {
		h := ch.handlers[strings.Join(args[:n], "/")]
		if h != nil {
			return h(p, c, header, args[n:]...)
		}
	}
	return ch.defaultHandler(p, c, header, args...)
}

func (p *Plugin) ExecuteCommand(c *plugin.Context, commandArgs *model.CommandArgs) (*model.CommandResponse, *model.AppError) {
	isAdmin, err := p.IsAuthorizedAdmin(commandArgs.UserId)
	if err != nil {
		return responsef("error occurred while authorizing the command: %v", err), nil
	}
	if !isAdmin {
		return responsef("`/autolink` commands can only be executed by a system administrator or `autolink` plugin admins."), nil
	}

	args := strings.Fields(commandArgs.Command)
	if len(args) == 0 || args[0] != autolinkCommand {
		return responsef(helpText), nil
	}

	return autolinkCommandHandler.Handle(p, c, commandArgs, args[1:]...), nil
}

func executeList(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	var links []autolink.Autolink
	var refs []int
	var err error

	if len(args) > 0 && (args[0] == optTemplate || args[0] == optPattern) {
		links, refs, err = searchLinkRefByTemplateOrPattern(p, header, args...)
	} else {
		links, refs, err = searchLinkRef(p, false, args...)
	}
	if err != nil {
		return responsef("%v", err)
	}

	text := ""
	if len(refs) > 0 {
		for _, i := range refs {
			text += links[i].ToMarkdown(i + 1)
		}
	} else {
		for i, l := range links {
			text += l.ToMarkdown(i + 1)
		}
	}
	return responsef(text)
}

func executeDelete(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	if len(args) != 1 {
		return responsef(helpText)
	}
	oldLinks, refs, err := searchLinkRef(p, true, args...)
	if err != nil {
		return responsef("%v", err)
	}
	n := refs[0]

	removed := oldLinks[n]
	newLinks := oldLinks[:n]
	if n+1 < len(oldLinks) {
		newLinks = append(newLinks, oldLinks[n+1:]...)
	}

	err = saveConfigLinks(p, newLinks)
	if err != nil {
		return responsef(err.Error())
	}

	return responsef("removed: \n%v", removed.ToMarkdown(0))
}

func executeSet(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	if len(args) < 3 {
		return responsef(helpText)
	}

	links, refs, err := searchLinkRef(p, true, args...)
	if err != nil {
		return responsef("%v", err)
	}
	l := &links[refs[0]]

	fieldName := args[1]
	restOfCommand := header.Command[len(autolinkCommand):] // "/autolink "
	restOfCommand = restOfCommand[strings.Index(restOfCommand, args[0])+len(args[0]):]
	restOfCommand = restOfCommand[strings.Index(restOfCommand, args[1])+len(args[1]):]
	value := strings.TrimSpace(restOfCommand)

	switch fieldName {
	case optName:
		l.Name = value
	case optPattern:
		l.Pattern = value
	case optTemplate:
		l.Template = value
	case optScope:
		l.Scope = args[2:]
	case optDisableNonWordPrefix:
		boolValue, e := parseBoolArg(value)
		if e != nil {
			return responsef("%v", e)
		}
		l.DisableNonWordPrefix = boolValue
	case optDisableNonWordSuffix:
		boolValue, e := parseBoolArg(value)
		if e != nil {
			return responsef("%v", e)
		}
		l.DisableNonWordSuffix = boolValue
	case optWordMatch:
		boolValue, e := parseBoolArg(value)
		if e != nil {
			return responsef("%v", e)
		}
		l.WordMatch = boolValue
	case optLogHits:
		boolValue, e := parseBoolArg(value)
		if e != nil {
			return responsef("%v", e)
		}
		l.LogHits = boolValue
	case optDisabled:
		boolValue, e := parseBoolArg(value)
		if e != nil {
			return responsef("%v", e)
		}
		l.Disabled = boolValue
	case optProcessBotPosts:
		boolValue, e := parseBoolArg(value)
		if e != nil {
			return responsef("%v", e)
		}
		l.ProcessBotPosts = boolValue
	default:
		return responsef("%q is not a supported field, must be one of %q", fieldName,
			[]string{optName, optDisabled, optPattern, optTemplate, optScope, optDisableNonWordPrefix, optDisableNonWordSuffix, optWordMatch, optLogHits, optProcessBotPosts})
	}

	err = saveConfigLinks(p, links)
	if err != nil {
		return responsef(err.Error())
	}

	ref := args[0]
	if l.Name != "" {
		ref = l.Name
	}
	return executeList(p, c, header, ref)
}

func executeTest(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	if len(args) < 2 {
		return responsef(helpText)
	}

	links, refs, err := searchLinkRef(p, false, args...)
	if err != nil {
		return responsef("%v", err)
	}

	restOfCommand := header.Command[len(autolinkCommand):] // "/autolink "
	restOfCommand = restOfCommand[strings.Index(restOfCommand, args[0])+len(args[0]):]
	orig := strings.TrimSpace(restOfCommand)
	out := fmt.Sprintf("- Original: `%s`\n", orig)

	for _, ref := range refs {
		l := links[ref]
		l.Disabled = false
		err = l.Compile()
		if err != nil {
			return responsef("failed to compile link %s: %v", l.DisplayName(), err)
		}
		replaced, _ := l.Replace(orig)
		if replaced == orig {
			out += fmt.Sprintf("- Link %s: _no change_\n", l.DisplayName())
		} else {
			out += fmt.Sprintf("- Link %s: changed to `%s`\n", l.DisplayName(), replaced)
			orig = replaced
		}
	}

	return responsef(out)
}

func executeEnable(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	if len(args) != 1 {
		return responsef(helpText)
	}
	return executeEnableImpl(p, c, header, args[0], true)
}

func executeDisable(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	if len(args) != 1 {
		return responsef(helpText)
	}
	return executeEnableImpl(p, c, header, args[0], false)
}

func executeEnableImpl(p *Plugin, c *plugin.Context, header *model.CommandArgs, ref string, enabled bool) *model.CommandResponse {
	links, refs, err := searchLinkRef(p, true, ref)
	if err != nil {
		return responsef("%v", err)
	}
	l := &links[refs[0]]
	l.Disabled = !enabled

	err = saveConfigLinks(p, links)
	if err != nil {
		return responsef(err.Error())
	}

	if l.Name != "" {
		ref = l.Name
	}
	return executeList(p, c, header, ref)
}

func executeAdd(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	if len(args) > 1 {
		return responsef(helpText)
	}
	name := ""
	if len(args) == 1 {
		name = args[0]
	}

	err := saveConfigLinks(p, append(p.getConfig().Links, autolink.Autolink{
		Name: name,
	}))
	if err != nil {
		return responsef(err.Error())
	}

	if name == "" {
		return executeList(p, c, header)
	}
	return executeList(p, c, header, name)
}

func executeHelp(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
	return responsef(helpText)
}

func responsef(format string, args ...interface{}) *model.CommandResponse {
	return &model.CommandResponse{
		ResponseType: model.CommandResponseTypeEphemeral,
		Text:         fmt.Sprintf(format, args...),
		Type:         model.PostTypeDefault,
	}
}

func searchLinkRef(p *Plugin, requireUnique bool, args ...string) ([]autolink.Autolink, []int, error) {
	links := p.getConfig().Sorted().Links
	if len(args) == 0 {
		if requireUnique {
			return nil, nil, errors.New("unreachable")
		}

		return links, nil, nil
	}

	n, err := strconv.ParseUint(args[0], 10, 32)
	if err == nil {
		if n < 1 || int(n) > len(links) {
			return nil, nil, errors.Errorf("%v is not a valid link number", n)
		}
		return links, []int{int(n) - 1}, nil
	}

	found := []int{}
	for i, l := range links {
		if strings.Contains(l.Name, args[0]) {
			found = append(found, i)
		}
	}
	if len(found) == 0 {
		return nil, nil, errors.Errorf("%q not found", args[0])
	}
	if requireUnique && len(found) > 1 {
		names := []string{}
		for _, i := range found {
			names = append(names, links[i].Name)
		}
		return nil, nil, errors.Errorf("%q matched more than one link: %q", args[0], names)
	}

	return links, found, nil
}

func searchLinkRefByTemplateOrPattern(p *Plugin, header *model.CommandArgs, args ...string) ([]autolink.Autolink, []int, error) {
	links := p.getConfig().Sorted().Links
	if len(args) == 1 {
		return links, nil, nil
	}

	restOfCommand := header.Command[len(autolinkCommand):]
	restOfCommand = restOfCommand[strings.Index(restOfCommand, args[0])+len(args[0]):]
	value := strings.TrimSpace(restOfCommand)

	found := []int{}
	for i, l := range links {
		if (args[0] == optTemplate && strings.Contains(strings.ToLower(l.Template), strings.ToLower(value))) ||
			(args[0] == optPattern && strings.Contains(strings.ToLower(l.Pattern), strings.ToLower(value))) {
			found = append(found, i)
		}
	}

	if len(found) == 0 {
		return nil, nil, errors.Errorf("%q not found", value)
	}

	return links, found, nil
}

func parseBoolArg(arg string) (bool, error) {
	switch strings.ToLower(arg) {
	case "true", "on":
		return true, nil
	case "false", "off":
		return false, nil
	}
	return false, errors.Errorf("Not a bool, %q", arg)
}

func saveConfigLinks(p *Plugin, links []autolink.Autolink) error {
	conf := p.getConfig()
	conf.Links = links

	configMap, err := p.getConfig().ToMap()
	if err != nil {
		return err
	}

	appErr := p.API.SavePluginConfig(configMap)
	if appErr != nil {
		return appErr
	}
	return nil
}
